p2p/pre2p/raintree/network.go

```diff
@@ -0,0 +1,207 @@
+	"time"
+
+	"github.com/pokt-network/pocket/shared/config"
+
+	typesPre2P "github.com/pokt-network/pocket/p2p/pre2p/types"
+	cryptoPocket "github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/types"
+
+	"google.golang.org/protobuf/proto"
+)
+
+var _ typesPre2P.Network = &rainTreeNetwork{}
+
+type rainTreeNetwork struct {
+	selfAddr cryptoPocket.Address
+	addrBook typesPre2P.AddrBook
+
+	// TECHDEBT(olshansky): Consider optimizing these away if possible.
+	// Helpers / abstractions around `addrBook` for simpler implementation through additional
+	// storage & pre-computation.
+	addrBookMap  typesPre2P.AddrBookMap
+	addrList     []string
+	maxNumLevels uint32
+
+	// TECHDEBT(drewsky): What should we use for de-duping messages within P2P?
+	mempool types.Mempool
+}
+
+func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesPre2P.AddrBook, config *config.Config) typesPre2P.Network {
+	n := &rainTreeNetwork{
+		selfAddr: addr,
+		addrBook: addrBook,
+		// This subset of fields are initialized by `handleAddrBookUpdates` below
+		addrBookMap:  make(typesPre2P.AddrBookMap),
+		addrList:     make([]string, 0),
+		maxNumLevels: 0,
+		// TODO(team): Mempool size should be configurable
+		mempool: types.NewMempool(1000000, 1000),
+	}
+
+	if err := n.handleAddrBookUpdates(); err != nil {
+		// DISCUSS(drewsky): if this errors, the node could still function but not participate in
+		// message propagation. Should we return an error or just log?
+		log.Println("[ERROR] Error initializing rainTreeNetwork: ", err)
+	}
+
+	return typesPre2P.Network(n)
+}
+
+func (n *rainTreeNetwork) NetworkBroadcast(data []byte) error {
+	return n.networkBroadcastInternal(data, n.maxNumLevels, getNonce())
+}
+
+func (n *rainTreeNetwork) networkBroadcastInternal(data []byte, level uint32, nonce uint64) error {
+	// This is handled either by the cleanup layer or redundancy layer
+	if level == 0 {
+		return nil
+	}
+
+	msg := &typesPre2P.RainTreeMessage{
+		Level: level,
+		Data:  data,
+		Nonce: nonce,
+	}
+	bz, err := proto.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	if addr1, ok := n.getFirstTargetAddr(level); ok {
+		if err = n.networkSendInternal(bz, addr1); err != nil {
+			log.Println("Error sending to peer during broadcast: ", err)
+		}
+	}
+
+	if addr2, ok := n.getSecondTargetAddr(level); ok {
+		if err = n.networkSendInternal(bz, addr2); err != nil {
+			log.Println("Error sending to peer during broadcast: ", err)
+		}
+	}
+
+	if err = n.demote(msg); err != nil {
+		log.Println("Error demoting self during RainTree message propagation: ", err)
+	}
+
+	return nil
+}
+
+func (n *rainTreeNetwork) demote(rainTreeMsg *typesPre2P.RainTreeMessage) error {
+	if rainTreeMsg.Level > 0 {
+		if err := n.networkBroadcastInternal(rainTreeMsg.Data, rainTreeMsg.Level-1, rainTreeMsg.Nonce); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *rainTreeNetwork) NetworkSend(data []byte, address cryptoPocket.Address) error {
```

Comment:
Same question about broadcast, can we do just `Send`?

Reply:
AGGREGATE